src/connection.rs

```diff
@@ -338,427 +387,495 @@ where
+
+                            // We ignore errors because sending can only fail if the `Handle`
+                            // has been dropped and if that happened, the `Drop` impl of
+                            // `Handle` has already triggered a close of this connection.
+                            let _ = admin.incoming.unbounded_send(stream);
+
+                            let stream = Box::new(s.map(sender::Message::from));
+                            let future = sender.send(sender::Message::AddStream(stream)).from_err()
+                                .map(move |sender| {
+                                    State::Ready(Connection::open(admin, input, sender))
+                                });
+                            return Box::new(future)
+                        }
 
-        Ok(None)
-    }
+                        // Data for an existing stream.
+                        match admin.streams.entry(id) {
+                            Entry::Occupied(entry) => {
+                                if frame.body().len() > entry.get().1.window() as usize {
+                                    error!("{}: {}: frame body too large", admin.id, id);
+                                    return immediate_close(admin, input, sender, ECODE_PROTO)
+                                }
+                                if is_finish {
+                                    entry.get().1.update_state(stream::State::WriteOnly);
+                                }
+                                let max_buffer_size = admin.config.max_buffer_size;
+                                // Stream buffer grows beyond limit => remove & reset stream
+                                if entry.get().1.buflen().map(move |n| n >= max_buffer_size).unwrap_or(true) {
+                                    warn!("{}: {}: max. stream buffer size reached", admin.id, id);
+                                    let stream = entry.remove();
+                                    stream.1.update_state(stream::State::Closed);
+                                    stream.1.notify_tasks();
+                                    let future = send_reset(id, sender)
+                                        .map(move |sender| {
+                                            State::Ready(Connection::open(admin, input, sender))
+                                        });
+                                    return Box::new(future)
+                                }
+
+                                let window = entry.get().1.decrement_window(frame.body().len() as u32);
+                                entry.get().1.add_data(frame.into_body());
+
+                                // If the stream window is closed, reset the window and grant the
+                                // remote more credit so it can keep sending data.
+                                if window == 0 && admin.config.window_update_mode == WindowUpdateMode::OnReceive {
```

Comment:
If the state of the stream is now `Closed`, which it may be if the stream was `ReadOnly` and this frame is a `FIN` (i.e. `is_finish` is true), then updating the window and sending a window update should probably  not be done?